src/KafkaFlow.Abstractions/Observer/Subject.cs

```diff
@@ -0,0 +1,53 @@
+namespace KafkaFlow.Observer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A generic implementation that should be extended to help the use of the notification system.
+    /// </summary>
+    /// <typeparam name="TSubject">The type of the subject.</typeparam>
+    /// <typeparam name="TArg">An argument type that will be passed to the observers</typeparam>
+    public abstract class Subject<TSubject, TArg> : ISubject<TSubject, TArg>
+        where TSubject : Subject<TSubject, TArg>
+    {
+        private readonly ILogHandler logHandler;
+        private readonly List<ISubjectObserver<TSubject, TArg>> observers = new();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Subject{TSubject, TArg}"/> class.
+        /// </summary>
+        /// <param name="logHandler">The log handler object to be used</param>
+        protected Subject(ILogHandler logHandler)
+        {
+            this.logHandler = logHandler;
+        }
+
+        /// <summary>
+        /// Subscribes an observer to the subject, allowing it to receive notifications.
+        /// </summary>
+        /// <param name="observer">The observer to subscribe.</param>
+        public void Subscribe(ISubjectObserver<TSubject, TArg> observer) => this.observers.Add(observer);
+
+        /// <summary>
+        /// Notifies all subscribed observers asynchronously.
+        /// </summary>
+        /// <param name="arg">The parameter passed by the client.</param>
+        /// <returns>A task representing the asynchronous notification operation.</returns>
+        public async Task NotifyAsync(TArg arg)
+        {
+            foreach (var observer in this.observers)
+            {
+                try
+                {
+                    await observer.OnNotification((TSubject)this, arg);
```

Comment:
A object that subscribes notifications that is disposed could lead to a NullReferenceException, correct?

Should this behavior result in a log error? we should probably do a null check instead for this use case, no?

Reply:
I like the unsubscribe method approach, it could be helpful in other scenarios

Reply:
I added the same suggestion in @ailtonguitar PR #438 - [(discussion here)](https://github.com/Farfetch/kafkaflow/pull/438#discussion_r1341956898)